metricbeat/module/vsphere/datastore/datastore.go

```diff
@@ -81,44 +102,140 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
-				},
-				"free": mapstr.M{
-					"bytes": ds.Summary.FreeSpace,
-				},
-				"used": mapstr.M{
-					"bytes": usedSpaceBytes,
-					"pct":   usedSpacePercent,
-				},
-			},
+	// Create a performance manager
+	perfManager := performance.NewManager(c)
+
+	// Retrieve all available metrics
+	metrics, err := perfManager.CounterInfoByName(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve metrics: %w", err)
+	}
+
+	// Filter for required metrics
+	var metricIds []types.PerfMetricId
+	for metricName := range metricSet {
+		if metric, ok := metrics[metricName]; ok {
+			metricIds = append(metricIds, types.PerfMetricId{CounterId: metric.Key})
+		} else {
+			m.Logger().Warnf("Metric %s not found", metricName)
 		}
+	}
 
-		reporter.Event(mb.Event{
-			MetricSetFields: event,
-		})
+	pc := property.DefaultCollector(c)
+	for i := range dst {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			assetNames, err := getAssetNames(ctx, pc, &dst[i])
+			if err != nil {
+				m.Logger().Errorf("Failed to retrieve object from host %s: %w", dst[i].Name, err)
+				continue
+			}
+
+			spec := types.PerfQuerySpec{
+				Entity:     dst[i].Reference(),
+				MetricId:   metricIds,
+				MaxSample:  1,
+				IntervalId: 20, // right now we are only grabbing real time metrics from the performance manager
+			}
+
+			// Query performance data
+			samples, err := perfManager.Query(ctx, []types.PerfQuerySpec{spec})
+			if err != nil {
+				m.Logger().Debugf("Failed to query performance data for host %s: %v", dst[i].Name, err)
+				continue
+			}
+
+			if len(samples) == 0 {
+				m.Logger().Debugf("No samples returned from performance manager")
+				continue
+			}
+
+			results, err := perfManager.ToMetricSeries(ctx, samples)
+			if err != nil {
+				m.Logger().Debugf("Failed to query performance data to metric series for host %s: %v", dst[i].Name, err)
+				continue
+			}
+
+			metricMap := make(map[string]interface{})
+			for _, result := range results[0].Value {
+				if len(result.Value) > 0 {
+					metricMap[result.Name] = result.Value[0]
+					continue
+				}
+				m.Logger().Debugf("For host %s,Metric %v: No result found", dst[i].Name, result.Name)
+			}
+
+			reporter.Event(mb.Event{
+				MetricSetFields: m.eventMapping(dst[i], &metricData{
+					perfMetrics: metricMap,
+					assetsNames: *assetNames,
+				}),
+			})
+		}
 	}
 
 	return nil
 }
+
+func getAssetNames(ctx context.Context, pc *property.Collector, ds *mo.Datastore) (*assetNames, error) {
+	referenceList := make([]types.ManagedObjectReference, 0, len(ds.Vm))
+	referenceList = append(referenceList, ds.Vm...)
+
+	var objects []mo.ManagedEntity
+	if len(referenceList) > 0 {
+		if err := pc.Retrieve(ctx, referenceList, []string{"name"}, &objects); err != nil {
+			return nil, err
+		}
+	}
+
+	outputVmNames := make([]string, 0, len(ds.Vm))
+	for _, ob := range objects {
+		if ob.Reference().Type == "VirtualMachine" {
+			name := strings.ReplaceAll(ob.Name, ".", "_")
```

Comment:
IMO, we should not change the names from "." to "_".
But as discussed we can take care of this as part of [this](https://github.com/elastic/obs-integration-team/issues/41) issue as this could be a breaking change for existing users and we will have to log that.